### PR TITLE
[Merged by Bors] - feat: port Data.List.Rotate

### PR DESCRIPTION
**Pull request by qawbecrdtey**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
I think this should wait for Data.Fin.Basic and then the `nthLe` lemmas can be restated with `get` where the statement is a bit nicer anyway.

File: Mathlib/Data/List/Rotate.lean [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+  simp [tsub_add_cancel_of_le, hi.le, hj.le, h]
+#align list.nodup.rotate_congr List.Nodup.rotate_congr
+
+section isRotated
+
+variable (l l' : List α)
+
+/-- `isRotated l₁ l₂` or `l₁ ~r l₂` asserts that `l₁` and `l₂` are cyclic permutations
+  of each other. This is defined by claiming that `∃ n, l.rotate n = l'`. -/
+def isRotated : Prop :=
```
Comment (jcommelin) [inline Mathlib/Data/List/Rotate.lean:387]:
```suggestion
def IsRotated : Prop :=
```

File: Mathlib/Data/List/Rotate.lean [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+  ∃ n, l.rotate n = l'
+#align list.is_rotated List.isRotated
+
+@[inherit_doc List.isRotated]
+infixr:1000 " ~r " => isRotated
+
+variable {l l'}
+
+@[refl]
+theorem isRotated.refl (l : List α) : l ~r l :=
```
Comment (jcommelin) [inline Mathlib/Data/List/Rotate.lean:397]: Please also fix all these

File: Mathlib/Data/List/Rotate.lean [hunk 3]
```diff
@@ -14,12 +14,12 @@ import Mathlib.Data.List.Range
...
 # List rotation
 
-This file proves basic results about `list.rotate`, the list rotation.
+This file proves basic results about `List.rotate`, the list rotation.
 
 ## Main declarations
 
-* `is_rotated l₁ l₂`: States that `l₁` is a rotated version of `l₂`.
-* `cyclic_permutations l`: The list of all cyclic permutants of `l`, up to the length of `l`.
+* `isRotated l₁ l₂`: States that `l₁` is a rotated version of `l₂`.
```
Comment (jcommelin) [inline Mathlib/Data/List/Rotate.lean:21]:
```suggestion
* `IsRotated l₁ l₂`: States that `l₁` is a rotated version of `l₂`.
```

**Review by jcommelin — COMMENTED** [review]
Otherwise, LGTM

File: Mathlib/Data/List/Rotate.lean [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+    · intro h
+      simpa using h.reverse
+#align list.is_rotated_reverse_comm_iff List.isRotated_reverse_comm_iff
+
+@[simp]
+theorem isRotated_reverse_iff : l.reverse ~r l'.reverse ↔ l ~r l' := by
+  simp [isRotated_reverse_comm_iff]
+#align list.is_rotated_reverse_iff List.isRotated_reverse_iff
+
+theorem is_rotated_iff_mod : l ~r l' ↔ ∃ n ≤ l.length, l.rotate n = l' := by
```
Comment (jcommelin) [inline Mathlib/Data/List/Rotate.lean:491]:
```suggestion
theorem isRotated_iff_mod : l ~r l' ↔ ∃ n ≤ l.length, l.rotate n = l' := by
```

File: Mathlib/Data/List/Rotate.lean [hunk 5]
```diff
@@ -0,0 +1,675 @@
...
+  refine' ⟨fun h => _, fun ⟨n, _, h⟩ => ⟨n, h⟩⟩
+  obtain ⟨n, rfl⟩ := h
+  cases' l with hd tl
+  · simp
+  · refine' ⟨n % (hd :: tl).length, _, rotate_mod _ _⟩
+    refine' (Nat.mod_lt _ _).le
+    simp
+#align list.is_rotated_iff_mod List.is_rotated_iff_mod
```
Comment (jcommelin) [inline Mathlib/Data/List/Rotate.lean:499]:
```suggestion
#align list.is_rotated_iff_mod List.isRotated_iff_mod
```

**Review by jcommelin — COMMENTED** [review]
bors d+

File: Mathlib/Data/List/Rotate.lean [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+  obtain ⟨n, rfl⟩ := h
+  cases' l with hd tl
+  · simp
+  · refine' ⟨n % (hd :: tl).length, _, rotate_mod _ _⟩
+    refine' (Nat.mod_lt _ _).le
+    simp
+#align list.is_rotated_iff_mod List.isRotated_iff_mod
+
+theorem isRotated_iff_mem_map_range : l ~r l' ↔ l' ∈ (List.range (l.length + 1)).map l.rotate := by
+  simp_rw [mem_map, mem_range, is_rotated_iff_mod]
```
Comment (qawbecrdtey) [inline Mathlib/Data/List/Rotate.lean:502]:
```suggestion
  simp_rw [mem_map, mem_range, isRotated_iff_mod]
```

**Review by qawbecrdtey — COMMENTED** [review]
Fixit

**Pull request closed without merging**